cmd/serve.go

```diff
@@ -24,18 +33,71 @@ Input event sources:
 func Serve(*cobra.Command, []string) {
 	config, err := configuration.AssembleConfig(viper.GetViper())
 	cobra.CheckErr(err)
-	client, err := pubsubapi.Client(config)
+
+	pubsubClient, err := pubsubapi.Client(config)
 	cobra.CheckErr(err)
+	pubsubCtx, cancelPubsub := context.WithCancel(context.Background())
+
+	apiServer := &http.Server{
+		Addr:    fmt.Sprintf(":%d", config.Api.Port),
+		Handler: api.NewRouter(config),
+	}
+
+	routines := []struct {
+		runForever   func()
+		uponShutdown func()
+	}{
+		{
+			runForever: func() {
+				shared.LogLn(config, "listening to pubsub...")
+				err := pubsub.ReceiveMessages(config, pubsubClient, pubsubCtx)
+				cobra.CheckErr(err)
+			},
+			uponShutdown: func() {
+				cancelPubsub()
+			},
+		},
+		{
+			runForever: func() {
+				shared.LogLn(config, fmt.Sprintf("serving api on port %d...", config.Api.Port))
+				if err := apiServer.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+					cobra.CheckErr(err)
+				}
+			},
+			uponShutdown: func() {
+				apiShutdownCtx, apiShutdownCancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer apiShutdownCancel()
+				err := apiServer.Shutdown(apiShutdownCtx)
+				cobra.CheckErr(err)
```

Comment:
should errors be logged here instead of `cobra.CheckErr`, which exits? Seems like this would kill other async tasks, eg. any pubsub processing that's wrapping up

Reply:
That totally slipped my mind, great catch. 

Rather than just logging, I switched from using waitgroup to using errgroup, which collects errors from all running coroutines and returns them when it is done (093cc5e). Means that we'll get the right exit code if there was an error in any of the shutdown routines.